pkg/app_gtwy/server/aether-app-gtwy-impl.go

```diff
@@ -0,0 +1,684 @@
+func (i *AppGtwy) GetDevices(ctx echo.Context, target externalRef0.Target, enterpriseID string, siteID string) error {
+
+	var listDeviceResponse []types.AppGtwyDevice
+
+	gnmiCtx, cancel := utils.NewGnmiContext(ctx, i.GnmiTimeout)
+	defer cancel()
+
+	site, err := i.gnmiGetEnterprisesEnterpriseSite(gnmiCtx, "/aether/v2.0.0/{target}/enterprises/enterprise/{enterprise-id}/site/{site-id}", target, enterpriseID, siteID)
+	if err != nil {
+		return utils.ConvertGrpcError(err)
+	}
+
+	si, err := i.AnalyticsClient.Query("subscriber_info")
+	if err != nil {
+		log.Errorf("error querying prometheus", err)
+		return err
+	}
+
+	simMap := mapSim(site)
+	deviceGroupMap := mapDeviceToDeviceGroups(site)
+	imsiMap := mapIMSI(si)
+
+	deviceMap := make(map[string]types.AppGtwyDevice)
+	for _, d := range *site.Device {
+		s := simMap[*d.SimCard]
+		newDev := types.AppGtwyDevice{
+			DeviceId:     &d.DeviceId,
+			DisplayName:  d.DisplayName,
+			Description:  d.Description,
+			Imei:         d.Imei,
+			SimIccid:     s.Iccid,
+			DeviceGroups: deviceGroupMap[d.DeviceId],
+		}
+		deviceMap[d.DeviceId] = newDev
+
+		imsi := strconv.Itoa(int(*s.Imsi))
+		if _, ok := imsiMap[imsi]; ok {
+			ip := string(imsiMap[imsi].Metric["ip"])
+			newDev.Ip = &ip
+			attached := imsiMap[imsi].Value.String()
+			newDev.Attached = &attached
+		}
+
+		listDeviceResponse = append(listDeviceResponse, newDev)
+	}
+
+	log.Infof("GetAppGtwyDevices")
+	return ctx.JSON(http.StatusOK, listDeviceResponse)
+}
+
+// GetDevice /appgtwy/v1/{target}/enterprises/{enterprise-id}/sites/{site-id}/devices/{device-id}
+func (i *AppGtwy) GetDevice(ctx echo.Context, target externalRef0.Target, enterpriseID string, siteID string, deviceId string) error {
+
+	gnmiCtx, cancel := utils.NewGnmiContext(ctx, i.GnmiTimeout)
+	defer cancel()
+
+	site, err := i.gnmiGetEnterprisesEnterpriseSite(gnmiCtx, "/aether/v2.0.0/{target}/enterprises/enterprise/{enterprise-id}/site/{site-id}", target, enterpriseID, siteID)
+	if err != nil {
+		return utils.ConvertGrpcError(err)
+	}
+
+	si, err := i.AnalyticsClient.Query("subscriber_info")
+	if err != nil {
+		return err
+	}
+
+	simMap := mapSim(site)
+	deviceGroupMap := mapDeviceToDeviceGroups(site)
+	imsiMap := mapIMSI(si)
+
+	var device types.AppGtwyDevice
+	for _, d := range *site.Device {
+		if d.DeviceId != deviceId {
+			continue
+		}
+		s := simMap[*d.SimCard]
+		deviceID := &d.DeviceId
+		device = types.AppGtwyDevice{
+			DeviceId:     deviceID,
+			DisplayName:  d.DisplayName,
+			Description:  d.Description,
+			Imei:         d.Imei,
+			SimIccid:     s.Iccid,
+			DeviceGroups: deviceGroupMap[deviceId],
+		}
+
+		imsi := strconv.Itoa(int(*s.Imsi))
+		if _, ok := imsiMap[imsi]; ok {
+			ip := string(imsiMap[imsi].Metric["ip"])
+			device.Ip = &ip
+			attached := imsiMap[imsi].Value.String()
+			device.Attached = &attached
+		}
+	}
+
+	log.Infof("GetDevice")
+	return ctx.JSON(http.StatusOK, device)
```

Comment:
I think if you return inside the loop, then you can conclude that if you get to here the device was not found and you can return an error